src/main/java/appeng/items/tools/powered/AbstractPortableCell.java

```diff
@@ -194,14 +197,82 @@ public long insert(Player player, ItemStack stack, AEKey what, AEKeyType allowed
+    public boolean overrideStackedOnOther(ItemStack stack, Slot slot, ClickAction action, Player player) {
+        if (action != ClickAction.SECONDARY || !slot.allowModification(player)) {
+            return false;
+        }
+
+        var other = slot.getItem();
+        if (other.isEmpty()) {
+            return true;
+        }
+
+        tryInsertFromPlayerOwnedItem(player, stack, other, getKeyType());
+        return true;
+    }
 
+    /**
+     * Allows directly inserting items and fluids into portable cells by right-clicking the cell with the item or bucket
+     * in hand.
+     */
     @Override
-    public abstract boolean overrideOtherStackedOnMe(ItemStack stack, ItemStack other, Slot slot, ClickAction action,
-            Player player, SlotAccess access);
+    public boolean overrideOtherStackedOnMe(ItemStack stack, ItemStack other, Slot slot, ClickAction action,
+            Player player, SlotAccess access) {
+        if (action != ClickAction.SECONDARY || !slot.allowModification(player)) {
+            return false;
+        }
+
+        if (other.isEmpty()) {
+            return false;
+        }
+
+        tryInsertFromPlayerOwnedItem(player, stack, other, getKeyType());
+        return true;
+    }
+
+    protected boolean tryInsertFromPlayerOwnedItem(Player player,
+            ItemStack cellStack,
+            ItemStack otherStack,
+            @Nullable AEKeyType keyType) {
+        if (keyType == null || keyType == AEKeyType.items()) {
+            AEKey key = AEItemKey.of(otherStack);
+            int inserted = (int) insert(player, cellStack, key, keyType, otherStack.getCount(), Actionable.MODULATE);
+            otherStack.shrink(inserted);
+            return inserted > 0;
+        } else {
+            var context = StackInteractions.findOwnedItemContext(keyType, player, otherStack);
+            if (context != null) {
+                var containedStack = context.getExtractableContent();
+                if (containedStack != null) {
+                    if (insert(player, cellStack, containedStack.what(), keyType, containedStack.amount(),
+                            Actionable.SIMULATE) == containedStack.amount()) {
+                        var extracted = context.extract(containedStack.what(), containedStack.amount(),
+                                Actionable.MODULATE);
+                        if (extracted > 0) {
+                            insert(player, cellStack, containedStack.what(), keyType, extracted, Actionable.MODULATE);
+                            context.playEmptySound(player, containedStack.what());
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @return The key-type allowed in this cell. Can be null to allow any.
+     */
+    @Nullable
+    protected abstract AEKeyType getKeyType();
```

Comment:
Maybe setting this to `null` by default so as to not need to define this for cells with custom transfer behaviour altogether? (e.g. cells with multiple supported types)

Reply:
Fair enough.